src/transformers/models/maskformer/modeling_maskformer.py

```diff
@@ -0,0 +1,1638 @@
+            m[: img.shape[1], : img.shape[2]] = False
+    else:
+        raise ValueError("not supported")
+    return NestedTensor(tensor, mask)
+
+
+# copied from original implementation
+def is_dist_avail_and_initialized():
+    if not dist.is_available():
+        return False
+    if not dist.is_initialized():
+        return False
+    return True
+
+
+# refactored from original implementation
+def dice_loss(inputs: Tensor, targets: Tensor, num_masks: float) -> Tensor:
+    r"""
+    Compute the DICE loss, similar to generalized IOU for masks as follow
+
+    .. math::
+        \mathcal{L}_{\text{dice}(x, y) = 1 - \frac{2 * x \cap y }{x \cup y + 1}}
+
+    In practice, since `targets` is a binary mask, (only 0s and 1s), dice can be computed as follow
+
+    .. math::
+        \mathcal{L}_{\text{dice}(x, y) = 1 - \frac{2 * x * y }{x + y + 1}}
+
+    Args:
+        inputs (Tensor): A tensor representing a mask
+        targets (Tensor): A tensor with the same shape as inputs. Stores the binary classification labels for each element in inputs (0 for the negative class and 1 for the positive class).
+
+
+    Returns:
+        Tensor: The computed loss
+    """
+    probs: Tensor = inputs.sigmoid().flatten(1)
+    numerator: Tensor = 2 * (probs * targets).sum(-1)
+    denominator: Tensor = probs.sum(-1) + targets.sum(-1)
+    loss: Tensor = 1 - (numerator + 1) / (denominator + 1)
+    loss = loss.sum() / num_masks
+    return loss
+
+
+# copied from original implementation
+def sigmoid_focal_loss(
+    inputs: Tensor, targets: Tensor, num_masks: int, alpha: float = 0.25, gamma: float = 2
+) -> Tensor:
+    r"""
+       Focal loss proposed in `Focal Loss for Dense Object Detection
+    <https://arxiv.org/abs/1708.02002>`_ originally used in RetinaNet. The loss is computed as follows
+
+    .. math::
+         \mathcal{L}_{\text{focal loss} = -(1 - p_t)^{\gamma}\log{(p_t)}
+
+    where :math:`CE(p_t) = -\log{(p_t)}}`, CE is the standard Cross Entropy Loss
+
+    Please refer to equation (1,2,3) of the paper for a better understanding.
+
+
+    Args:
+        inputs (Tensor): A float tensor of arbitrary shape.
+                The predictions for each example.
+        targets (Tensor,): A tensor with the same shape as inputs. Stores the binary classification labels for each element in inputs (0 for the negative class and 1 for the positive class).
+        alpha (float, optional): Weighting factor in range (0,1) to balance
+                positive vs negative examples. Default = -1 (no weighting).
+        gamma (float, optional): Exponent of the modulating factor (1 - p_t) to
+                balance easy vs hard examples.
+    Returns:
+        Tensor: The computed loss
+    """
+    probs: Tensor = inputs.sigmoid()
+    ce_loss: Tensor = F.binary_cross_entropy_with_logits(inputs, targets, reduction="none")
+    p_t: Tensor = probs * targets + (1 - probs) * (1 - targets)
+    loss: Tensor = ce_loss * ((1 - p_t) ** gamma)
```

Comment:
The `: Tensor` hurt readbility IMO. In a modeling file, pretty much everything is a tensor, so it's not an interesting type to mark.
`p_t` is too short a name, as is `ce_loss`. It would be more interesting to have longer names and be able to follow what those things represent IMO.

Reply:
`p_t` comes from the paper equation, that is written in the doc of the function. I'd like to keep that name since it's consistent